### PR TITLE
Certificate Based Authentication

### PR DESCRIPTION
**Pull request by landenblackmsft**
Enable Certificate Based Authentication.

Add support for user input --client_certificate_path
If available, add certfile to context via load_cert_chain.

### CODE REVIEW

File: build-restler.py [hunk 1]
```diff
@@ -160,7 +160,7 @@ def publish_dotnet_apps(dirs, configuration):
         proj_file_path = dotnetcore_projects[target_dir_name]
         print(f"Publishing project {proj_file_path} to output dir {proj_output_dir}")
 
-        output = subprocess.run(f"dotnet restore {proj_file_path} --use-lock-file --locked-mode --force", shell=True, stderr=subprocess.PIPE)
+        output = subprocess.run(f"dotnet restore {proj_file_path} --use-lock-file --locked-mode --force -s https://api.nuget.org/v3/index.json", shell=True, stderr=subprocess.PIPE)
         if output.stderr:
```
Comment (marina-p) [inline build-restler.py:164]: Could you please either delete this before merging, or add a "--nuget_source" argument and pass it here?

File: src/driver/Program.fs [hunk 2]
```diff
@@ -515,6 +516,11 @@ let rec parseEngineArgs task (args:EngineParameters) = function
         parseEngineArgs task { args with checkerOptions = args.checkerOptions @ [(checkerAction, specifiedCheckers |> String.concat " ")] } rest
     | "--no_results_analyzer"::rest ->
         parseEngineArgs task { args with runResultsAnalyzer = false } rest
+    | "--client_certificate_path"::certFilePath::rest ->
+        if not (File.Exists certFilePath) then
+            Logging.logError <| sprintf "The RESTler client certificate file path %s does not exist." certFilePath
```
Comment (marina-p) [inline src/driver/Program.fs:521]: >RESTler [](http://example.com/codeflow?start=45&length=7)

remove 'RESTler'.

File: src/driver/Types.fs [hunk 3]
```diff
@@ -65,6 +65,8 @@ type EngineParameters =
         /// Specifies whether to run results analyzer.
         runResultsAnalyzer : bool
 
+        /// File path to the client certificate.
+        certFilePath : string
```
Comment (marina-p) [inline src/driver/Types.fs:69]: >string [](http://example.com/codeflow?start=23&length=6)

Could you please make this a string option, and set the default to None below?  Then, in Program.fs you can check for the None value so the parameter isn't sent to the engine at all when the user does not provide a cert.

File: src/driver/Program.fs [hunk 4]
```diff
@@ -208,6 +208,7 @@ module Fuzz =
             sprintf "--restler_grammar \"%s\"" parameters.grammarFilePath
             sprintf "--custom_mutations \"%s\"" parameters.mutationsFilePath
             sprintf "--set_version %s" CurrentVersion
+            sprintf "--client_certificate_path \"%s\"" parameters.certFilePath
 
             (match parameters.refreshableTokenOptions with
```
Comment (marina-p) [inline src/driver/Program.fs:213]: >refreshableTokenOptions [](http://example.com/codeflow?start=30&length=23)

See comment in Types.fs - could you please make this parameter only be passed if the user sets the option?  

match parameters.certFilePath with
| None -> ""
| Some cfPath -> <your current sprintf>

**Review by marina-p — CHANGES_REQUESTED** [review]
:clock1:

File: docs/user-guide/SettingsFile.md [hunk 5]
```diff
@@ -27,6 +27,11 @@ Checkers' Friendly Names:
 * PayloadBody
 * Examples
 
+### client_certificate_path: str (default None)
+Path to your X.509 certificate file in PEM format.
+
+If provided and valid, RESTler will attempt to use it during the SSL handshake.
```
Comment (landenblackmsft) [inline docs/user-guide/SettingsFile.md:33]: Is this reference to "RESTler" okay, or should it be changed?

Comment (marina-p) [reply docs/user-guide/SettingsFile.md:33]: Yes, this looks good.

**Review by marina-p — APPROVED** [review]
:shipit:

**Comment by marina-p** [comment]
Thanks for contributing this feature!

**Review by marina-p — APPROVED** [review]
:shipit:

**Pull request merged**